### PR TITLE
Delegate SerializableStackTraceElement.toString() to Java's own

### PR DESCRIPTION
**Pull request by a-k-g**
#347

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/349)

### CODE REVIEW

File: error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java [hunk 1]
```diff
@@ -30,13 +30,23 @@
 @JsonIgnoreProperties(ignoreUnknown = true)
 public abstract class SerializableStackTraceElement implements Serializable {
 
-    public abstract Optional<String> getClassName();
+    public abstract String getClassName();
```
Comment (a-k-g) [inline error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java]: Not sure if we really needed these `Optionals` as `StackTraceElement` has only the following constructor-
```
    public StackTraceElement(String declaringClass, String methodName,
                             String fileName, int lineNumber) {
        this.declaringClass = Objects.requireNonNull(declaringClass, "Declaring class is null");
        this.methodName     = Objects.requireNonNull(methodName, "Method name is null");
        this.fileName       = fileName;
        this.lineNumber     = lineNumber;
    }
```
And given that we're 2.0-ing it seems like a good opportunity to make this breaking change.

File: error-handling/src/main/java/com/palantir/remoting2/errors/SerializableError.java [hunk 2]
```diff
@@ -92,10 +92,10 @@ public static SerializableError of(
                         @Override
                         public SerializableStackTraceElement apply(@Nullable StackTraceElement input) {
                             return SerializableStackTraceElement.builder()
-                                    .className(Optional.ofNullable(input.getClassName()))
-                                    .methodName(Optional.ofNullable(input.getMethodName()))
+                                    .className(input.getClassName())
```
Comment (uschi2000) [inline error-handling/src/main/java/com/palantir/remoting2/errors/SerializableError.java:95]: cheers, this was silly

File: error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java (hunk 1 above)
Comment (uschi2000) [reply error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java]: this isn't necessarily coming from java, so we need to keep the optionality in the wire format

File: error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java [hunk 3]
```diff
@@ -30,13 +30,23 @@
...
+    public abstract String getMethodName();
 
     public abstract Optional<String> getFileName();
 
-    public abstract Optional<Integer> getLineNumber();
+    public abstract int getLineNumber();
+
+    @Override
+    public final String toString() {
+        StackTraceElement element = new StackTraceElement(getClassName(),
```
Comment (uschi2000) [inline error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java]: cannot guarantee that these are non-null, so cannot simply construct this. IMO, it's reasonable to do:
```
StackTraceElement element = new StackTraceElement(
  getClassName().orElse("UnknownClass"),
  getMethodName().orElse("unknownMethod"),
  getFileName.orElse(null),
  getLineNumer.orElse(0));
```

Comment (uschi2000) [reply error-handling/src/main/java/com/palantir/remoting2/errors/SerializableStackTraceElement.java]: please add a test verifying that this works when class and method are invalid java identifiers, i.e., 'Bogus Class Name', and 'bogus method name'.

**Review by uschi2000 — CHANGES_REQUESTED** [review]
few changes, see above

**Review by uschi2000 — APPROVED** [review]
cheers!

**Pull request merged**